File: frontend/types/assignment.ts
export interface AssignmentListItem {
  id: string;
  title: string;
  subject: string;
  description?: string;
  due_date: string;
  max_score: number;
  submission_count: number;
  graded_count: number;
  created_at: string;
}

export interface Assignment extends AssignmentListItem {
  class_id: string;
  class_name: string;
  file_url?: string;
}

export interface CreateAssignmentRequest {
  title: string;
  subject: string;
  description?: string;
  due_date: string;
  max_score: number;
  file_url?: string;
}

export interface AssignmentSubmission {
  id: string;
  student_id: string;
  student_name: string;
  submitted_at: string;
  file_url?: string;
  score?: number;
  feedback?: string;
  is_graded: boolean;
}

export interface GradeSubmissionRequest {
  score: number;
  feedback?: string;
}

export interface AssignmentListParams {
  page?: number;
  per_page?: number;
  subject?: string;
}

File: backend/app/schemas/assignment.py
import uuid
from datetime import date, datetime
from typing import Optional

from pydantic import BaseModel, Field


# ---------------------------------------------------------------------------
# Assignment requests
# ---------------------------------------------------------------------------

class AssignmentCreate(BaseModel):
    class_id: uuid.UUID
    title: str = Field(..., min_length=1, max_length=255)
    subject: str = Field(..., min_length=1, max_length=100)
    description: Optional[str] = None
    due_date: date
    max_score: float = Field(100.0, gt=0)
    file_url: Optional[str] = Field(None, max_length=500)


class AssignmentUpdate(BaseModel):
    title: Optional[str] = Field(None, min_length=1, max_length=255)
    subject: Optional[str] = Field(None, min_length=1, max_length=100)
    description: Optional[str] = None
    due_date: Optional[date] = None
    max_score: Optional[float] = Field(None, gt=0)
    file_url: Optional[str] = Field(None, max_length=500)


# ---------------------------------------------------------------------------
# Assignment responses
# ---------------------------------------------------------------------------

class AssignmentResponse(BaseModel):
    id: uuid.UUID
    class_id: uuid.UUID
    class_name: str
    teacher_name: str
    title: str
    subject: str
    description: Optional[str]
    due_date: date
    max_score: float
    file_url: Optional[str]
    submission_count: int
    graded_count: int
    created_at: datetime

    model_config = {"from_attributes": True}


class PaginatedAssignmentResponse(BaseModel):
    total: int
    page: int
    per_page: int
    items: list[AssignmentResponse]


# ---------------------------------------------------------------------------
# Submission
# ---------------------------------------------------------------------------

class SubmissionResponse(BaseModel):
    id: uuid.UUID
    assignment_id: uuid.UUID
    assignment_title: str
    student_id: uuid.UUID
    student_name: str
    score: Optional[float]
    feedback: Optional[str]
    file_url: Optional[str]
    submitted_at: datetime
    graded_at: Optional[datetime]
    is_graded: bool

    model_config = {"from_attributes": True}


class GradeSubmissionRequest(BaseModel):
    score: float = Field(..., ge=0)
    feedback: Optional[str] = Field(None, max_length=1000)
